packages/core/src/codewhisperer/activation.ts

```diff
@@ -267,6 +274,93 @@ export async function activate(context: ExtContext): Promise<void> {
         await notifyNewCustomizations()
     }
 
+    function setSubscriptionsForAutoScans() {
+        // Initial scan when the editor opens for the first time
+        const editor = vscode.window.activeTextEditor
+        if (
+            CodeScansState.instance.isScansEnabled() &&
+            editor &&
+            securityScanLanguageContext.isLanguageSupported(editor.document.languageId) &&
+            editor.document.getText().length > 0
+        ) {
+            void debounceStartSecurityScan(
+                securityPanelViewProvider,
+                editor,
+                client,
+                context.extensionContext,
+                CodeWhispererConstants.SecurityScanType.File
+            )
+        }
+
+        context.extensionContext.subscriptions.push(
+            // Trigger scan if focus switches to a different file
+            vscode.window.onDidChangeActiveTextEditor(editor => {
+                if (
+                    CodeScansState.instance.isScansEnabled() &&
+                    editor &&
+                    securityScanLanguageContext.isLanguageSupported(editor.document.languageId)
+                ) {
+                    const codewhispererDiagnostics = securityScanRender.securityDiagnosticCollection
+                        ?.get(editor.document.uri)
+                        ?.filter(({ source }) => source === CodeWhispererConstants.codewhispererDiagnosticSourceLabel)
+
+                    if (
+                        (codewhispererDiagnostics && codewhispererDiagnostics.length > 0) ||
+                        editor.document.getText().length === 0
+                    ) {
+                        // Do nothing since we already have diagnostics for this file, or the file contents are empty.
+                    } else {
+                        void debounceStartSecurityScan(
+                            securityPanelViewProvider,
+                            editor,
+                            client,
+                            context.extensionContext,
+                            CodeWhispererConstants.SecurityScanType.File
+                        )
+                    }
+                }
+            }),
+            // Trigger scan if the file contents change
+            vscode.workspace.onDidChangeTextDocument(async event => {
+                const editor = vscode.window.activeTextEditor
+                if (
+                    CodeScansState.instance.isScansEnabled() &&
+                    editor &&
+                    event.document === editor.document &&
+                    securityScanLanguageContext.isLanguageSupported(editor.document.languageId) &&
+                    event.contentChanges.length > 0
```

Comment:
Are we going to trigger scans even for 1 character change? even though they are just file scans, wouldn't that be expensive?

Reply:
For now yes, we don't know whether the 1 character change will complete the function or not. We can adjust later once we get actual usage data